src/main/java/org/jitsi/videobridge/cc/AdaptiveTrackProjection.java

```diff
@@ -0,0 +1,307 @@
+     * @return true if the packet is accepted, false otherwise.
+     */
+    public boolean accept(@NotNull RawPacket rtpPacket)
+    {
+        AdaptiveTrackProjectionContext
+            contextCopy = getContext(RawPacket.getPayloadType(rtpPacket));
+
+        int targetIndexCopy = targetIndex;
+        boolean accept = contextCopy.accept(rtpPacket, targetIndexCopy);
+
+        if (contextCopy.needsKeyframe() && targetIndexCopy > -1)
+        {
+            MediaStreamTrackDesc source = getSource();
+            if (source != null)
+            {
+                ((RTPTranslatorImpl) source
+                    .getMediaStreamTrackReceiver()
+                    .getStream()
+                    .getRTPTranslator())
+                    .getRtcpFeedbackMessageSender()
+                    .requestKeyframe(targetSsrc);
+            }
+        }
+
+        return accept;
+    }
+
+    /**
+     * Gets or creates the adaptive track projection context that corresponds to
+     * the payload type that is specified as a parameter. If the payload type
+     * is different from {@link #contextPayloadType}, then a new adaptive track
+     * projection context is created that is appropriate for the new payload
+     * type.
+     *
+     * @param payloadType the payload type of the adaptive track projection to
+     * get or create.
+     * @return the adaptive track projection context that corresponds to
+     * the payload type that is specified as a parameter.
+     */
+    private synchronized
+    AdaptiveTrackProjectionContext getContext(int payloadType)
+    {
+        if (context == null || contextPayloadType != payloadType)
+        {
+            MediaStreamTrackDesc source = getSource();
+            MediaFormat format = source
+                .getMediaStreamTrackReceiver()
+                .getStream().getDynamicRTPPayloadTypes().get((byte) payloadType);
+
+            context = makeContext(source, format);
+            contextPayloadType = payloadType;
+            return context;
+        }
+        else
+        {
+            return context;
+        }
+    }
+
+    /**
+     * Utility/factory method that creates the appropriate adaptive track
+     * projection context based on the media format that is specified as a
+     * parameter. Note that a media format is the same thing as a payload type
+     * (there's a one-to-one mapping between payload type and media format).
+     *
+     * @param track the ssrc of the track projection.
+     * @param format the media format.
+     * @return an adaptive track projection context that corresponds to the
+     * media format that is specified as a parameter.
+     */
+    private static AdaptiveTrackProjectionContext makeContext(
+        @NotNull MediaStreamTrackDesc track, @NotNull MediaFormat format)
+    {
+        if (Constants.VP8.equalsIgnoreCase(format.getEncoding())
+            && track.getRTPEncodings().length > 1)
```

Comment:
can use `!track.getRTPEncodings().isEmpty())`

Reply:
With non-adaptive tracks we have a single encoding/quality (the array is non-empty). Here we want to distinguish VP8 simulcast+svc (the array has 9 encodings/qualities) vs anything else.

Reply:
ah sorry!  i misread the `>1`, you're right. my bad.